virtualmachine/vsphere/vm.go

```diff
@@ -1678,49 +1690,104 @@ func CreateTemplate(vm *VM) error {
 	return err
 }
 
-// GetTemplateList : Returns the template VMs in a cluster
-func GetTemplateList(vm *VM) ([]map[string]interface{}, error) {
+// getOsDetails: returns details of guest os
+func getOsDetails(vmMo mo.VirtualMachine) map[string]interface{} {
+	osDetails := make(map[string]interface{})
+	if vmMo.Guest == nil {
+		return osDetails
+	}
+	osDetails = map[string]interface{}{
+		"guest_id":        vmMo.Guest.GuestId,
+		"guest_family":    vmMo.Guest.GuestFamily,
+		"guest_full_name": vmMo.Guest.GuestFullName,
+		"hostname":        vmMo.Guest.HostName,
+	}
+	return osDetails
+}
+
+// getVmInfo : get vm info from vm managed object
+func getVmInfo(vmMo mo.VirtualMachine, vmPath string) map[string]interface{} {
+	var (
+		toolsStatus bool
+	)
+	// fetching fisk info
+	diskInfo := make([]map[string]interface{}, 0)
+	for _, device := range vmMo.Config.Hardware.Device {
+		disk, ok := device.(*types.VirtualDisk)
+		if !ok {
+			continue
+		}
+		devinfo, ok := disk.DeviceInfo.(*types.Description)
+		if !ok {
+			continue
+		}
+		backing := disk.Backing.(types.BaseVirtualDeviceFileBackingInfo)
+		fileBackingInfo := backing.GetVirtualDeviceFileBackingInfo()
+		diskInfo = append(diskInfo, map[string]interface{}{
+			"name":      devinfo.Label,
+			"size":      disk.CapacityInKB,
+			"disk_file": fileBackingInfo.FileName,
+		})
+	}
+	if vmMo.Guest != nil {
+		toolsStatus = getToolsRunningStatus(
+			vmMo.Guest.ToolsRunningStatus)
+	}
+
+	return map[string]interface{}{
+		"name":           vmPath, // full name/path of vm
+		"id":             vmMo.Self.Value,
+		"instance_uuid":  vmMo.Summary.Config.InstanceUuid,
+		"memory_size_mb": vmMo.Summary.Config.MemorySizeMB,
+		"num_cpu":        vmMo.Summary.Config.NumCpu,
+		"disks":          diskInfo,
+		"ip_addresses":   getIpFromVmMo(&vmMo),
```

Comment:
Can any of below get calls fail?
If nil return is the failure case, how is the handled by callers?

Reply:
changed to returning empty slices instead of nil.